runtime/src/main/java/io/quarkiverse/logging/cloudwatch/LoggingCloudWatchHandler.java

```diff
@@ -97,49 +99,35 @@ public void close() throws SecurityException {
 
         @Override
         public void run() {
-            while (true) {
-                synchronized (eventBuffer) {
-                    events = new ArrayList<>(eventBuffer);
-                    eventBuffer.clear();
-                }
-
-                if (events.size() > 0) {
-                    PutLogEventsRequest request = new PutLogEventsRequest();
-                    request.setLogEvents(events);
-                    request.setLogGroupName(logGroupName);
-                    request.setLogStreamName(logStreamName);
+            synchronized (eventBuffer) {
+                events = new ArrayList<>(eventBuffer);
+                eventBuffer.clear();
+            }
+            boolean workingSequenceToken = false;
+            if (events.size() > 0) {
+                PutLogEventsRequest request = new PutLogEventsRequest();
+                request.setLogEvents(events);
+                request.setLogGroupName(logGroupName);
+                request.setLogStreamName(logStreamName);
+
+                while (!workingSequenceToken) {
                     request.setSequenceToken(sequenceToken);
-                    // Do the call and get the next token
 
                     try {
                         sequenceToken = awsLogs.putLogEvents(request).getNextSequenceToken();
+                        workingSequenceToken = true;
                     } catch (InvalidSequenceTokenException e) {
                         String exceptionMessage = e.getMessage();
                         LOGGER.info("exception message: " + exceptionMessage);
-                        String validSequenceToken = extractValidSequenceToken(exceptionMessage);
-                        sequenceToken = validSequenceToken;
-                        LOGGER.info("valid sequence token: " + validSequenceToken);
-                        LOGGER.info("actual sequence token: " + sequenceToken);
-
-                        PutLogEventsRequest newRequest = new PutLogEventsRequest();
-                        newRequest.setLogEvents(events);
-                        newRequest.setLogGroupName(logGroupName);
-                        newRequest.setLogStreamName(logStreamName);
-                        newRequest.setSequenceToken(validSequenceToken);
-
-                        sequenceToken = awsLogs.putLogEvents(newRequest).getNextSequenceToken();
+                        sequenceToken = extractValidSequenceToken(exceptionMessage);
+                        LOGGER.info("extracted sequence token: " + sequenceToken);
+                        workingSequenceToken = false;
                     }
                 }
-
-                if (done) {
-                    return;
-                }
-
-                try {
-                    Thread.sleep(5000);
-                } catch (InterruptedException e) {
-                    e.printStackTrace();
-                }
+            }
+            if (done) {
+                scheduler.shutdown();
```

Comment:
The scheduler shut down should happen when the server is shut down.

Quarkus has the [ShutdownEvent](https://quarkus.io/guides/lifecycle#listening-for-startup-and-shutdown-events) for that and if it works, that's probably the best time to do the shut down. I'm not sure it will work from a handler though.

The other possibility is to rely on the [Handler#close()](https://docs.oracle.com/en/java/javase/11/docs/api/java.logging/java/util/logging/Handler.html#close()) method call and shut down the executor at that time.

Reply:
Regarding the executor shutdown itself, the [ExecutorService Javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ExecutorService.html) has a nice example about how it should be done (scroll down to `shutdownAndAwaitTermination`).

I think we need to push all entries from `eventBuffer` to CW during the first phase of the shutdown.

Reply:
Thanks for the hint regarding the two step shutdown. That's great :+1:

Reply:
I prefer the close() solution.